src/core/namespace.c

```diff
@@ -1035,17 +1044,15 @@ static int make_read_only(const MountEntry *m, char **blacklist, FILE *proc_self
 
         if (mount_entry_read_only(m)) {
                 if (IN_SET(m->mode, EMPTY_DIR, TMPFS)) {
-                        /* Make superblock readonly */
-                        if (mount(NULL, mount_entry_path(m), NULL, MS_REMOUNT | MS_RDONLY | m->flags, NULL) < 0)
-                                r = -errno;
+                        r = remount_bind_readonly(mount_entry_path(m), m->flags);
                 } else {
                         submounts = true;
                         r = bind_remount_recursive_with_mountinfo(mount_entry_path(m), true, blacklist, proc_self_mountinfo);
                 }
         } else if (m->mode == PRIVATE_DEV) {
-                /* Superblock can be readonly but the submounts can't */
-                if (mount(NULL, mount_entry_path(m), NULL, MS_REMOUNT | MS_RDONLY | DEV_MOUNT_OPTIONS, NULL) < 0)
-                        r = -errno;
+                /* /dev can be readonly but the submounts can't.  Also, we only set the per-mount read-only flag.
+                 * We can't set it on the superblock, if we are inside a user namespace and running Linux <= 4.17. */
+                r = remount_bind_readonly(mount_entry_path(m), DEV_MOUNT_OPTIONS);
```

Comment:
Hmm, that does solve the problem I had hitting the 120 limit for line length...

Comparing your suggestion, I think my commit has more consistent behaviour.  It will give the same behaviour for `ReadOnlyPaths=/dev`, for either setting of `PrivateDevices=`.  This is the same behaviour we have before this commit, except for the superblock readonly flag.  I've written a new test to show this case.  I hope that makes it a bit more acceptable.

Thanks for reviewing my code, this is all great feedback.